homeassistant/components/flexmeasures/services.py

```diff
@@ -0,0 +1,178 @@
+        "service": SERVICE_CHANGE_CONTROL_TYPE,
+        "service_func_name": "change_control_type",
+    },
+    {
+        "schema": None,
+        "service": "trigger_and_get_schedule",
+        "service_func_name": "trigger_and_get_schedule",
+    },
+    {
+        "schema": None,
+        "service": "post_measurements",
+        "service_func_name": "post_measurements",
+    },
+]
+
+LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_services(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Set up services."""
+
+    # Is this the correct way and place to set this?
+    client: FlexMeasuresClient = hass.data[DOMAIN]["fm_client"]
+
+    ############
+    # Services #
+    ############
+
+    async def change_control_type(
+        call: ServiceCall,
+    ):  # pylint: disable=possibly-unused-variable
+        """Change control type S2 Protocol."""
+
+        if "cem" not in hass.data[DOMAIN]:
+            raise UndefinedCEMError()
+
+        cem: CEM = hass.data[DOMAIN]["cem"]
+
+        control_type = cast(str, call.data.get("control_type"))
+
+        if not hasattr(ControlType, control_type):
+            raise UnknownControlType()
+
+        control_type = ControlType[control_type]
+
+        await cem.activate_control_type(control_type=control_type)
+
+        hass.states.async_set(
+            f"{DOMAIN}.cem", json.dumps({"control_type": str(cem.control_type)})
+        )
+
+    async def trigger_and_get_schedule(
+        call: ServiceCall,
+    ):  # pylint: disable=possibly-unused-variable
+        resolution = pd.Timedelta(RESOLUTION)
+        tzinfo = dt_util.get_time_zone(hass.config.time_zone)
+        start = time_ceil(datetime.now(tz=tzinfo), resolution)
+
+        flex_model = client.create_storage_flex_model(
+            soc_at_start=call.data.get("soc_at_start"),
+            soc_unit=SOC_UNIT,
+            soc_max=get_from_option_or_config("soc_max", entry),
+            soc_min=get_from_option_or_config("soc_min", entry),
+            soc_targets=call.data.get("soc_targets"),
+        )
+        flex_context = client.create_storage_flex_context(
+            consumption_price_sensor=get_from_option_or_config(
+                "consumption_price_sensor", entry
+            ),
+            production_price_sensor=get_from_option_or_config(
+                "production_price_sensor", entry
+            ),
+        )
+        schedule_input = {
+            "sensor_id": get_from_option_or_config("power_sensor", entry),
+            "start": start,
+            "duration": get_from_option_or_config("schedule_duration", entry),
+            "flex_model": flex_model,
+            "flex_context": flex_context,
+        }
+        LOGGER.info(schedule_input)
+        schedule = await client.trigger_and_get_schedule(**schedule_input)
+
+        schedule = [
+            {"start": start + resolution * i, "value": value}
+            for i, value in enumerate(schedule["values"])
+        ]
```

Comment:
I don't think we should add variables that will not be used for the API call to the client functions. I've made a pull request to the fm client to add the unit conversion as a static method: https://github.com/FlexMeasures/flexmeasures-client/pull/68

In the integration the return values can be converted using this method.

Reply:
Okay. But just to note that, at some point, the API *will* accept the unit field. I think it is even in the API documentation already (but it might be broken in API v3, whereas older API versions have been sunset).